File: leetcode/1798-max-number-of-k-sum-pairs/solution.cpp
class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
    //     sort(nums.begin(), nums.end());
    //     int l = 0;
    //     int r = nums.size() - 1;
    //     int count = 0;
    //     while(l < r) {
    //         if (nums[l] + nums[r] == k) {
    //             l++;
    //             r--;
    //             count++;
    //         }
    //         else if (nums[l] + nums[r] > k) r--;
    //         else if (nums[l] + nums[r] < k) l++;
    //     }
    //     return count;
    
        unordered_map<int, int> hmap;
        int res = 0;
        for (int x: nums) {
            if (hmap[k-x] !=0) {
                hmap[k-x]--;
                res++;
            }
            else {
                hmap[x]++;
            }
        }
        return res;
    }
    
};

File: leetcode/1798-max-number-of-k-sum-pairs/solution.py
class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        # from collections import Counter
        # hmap = Counter(nums)
        # res = 0
        # for key in hmap.keys():
        #     #if not exist a pair, continue
        #     if k - key in hmap:
        #     #if exist a pair, use up
        #         while hmap[key]:
        #             hmap[key] -=1
        #             if hmap[k-key] == 0:
        #                 hmap[key] +=1
        #                 break
        #             else:
        #                 hmap[k-key] -=1
        #                 res +=1
        # return res
        hmap = {}
        res = 0
        for num in nums:
            if k - num in hmap and hmap[k-num] != 0:
                hmap[k-num]-=1
                res +=1
            else:
                if num in hmap:
                    hmap[num] += 1
                else:
                    hmap[num] = 1

        return res
